Fixed and better documented the is_trivially_moveable and moved_out_v…

psiha opened the pull request:
…alue_is_trivially_destructible traits.

Fixes https://github.com/psiha/vm/issues/32.
Clarifies https://github.com/psiha/vm/issues/31.

psiha committed:
Fixed and better documented the is_trivially_moveable and moved_out_value_is_trivially_destructible traits.
Fixes https://github.com/psiha/vm/issues/32.
Clarifies https://github.com/psiha/vm/issues/31.

Quuxplusone commented on include/psi/vm/containers/vector_impl.hpp:
```diff
@@ -127,26 +129,42 @@ class [[ nodiscard, clang::trivial_abi ]] vector_impl
+    // the stray contrived type that might not be included otherwise.
+    // (crossing ways with https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2839r0.html)
+    // TODO impl https://stackoverflow.com/questions/51901837/how-to-get-if-a-type-is-truly-move-constructible/51912859#51912859
+    static constexpr bool moved_out_value_is_trivially_destructible{ std::is_nothrow_move_assignable_v<T> || std::is_trivially_destructible_v<T> };
```
I'll come back with more comments later; for now, I'll just say that this strikes me as wrong.
For example, MSVC's `std::list<int>` is nothrow-move-assignable but not trivially-destructible-after-move. (Skipping the destructor will leak the sentinel node.) Another classic (if contrived) example is `cxx03_shared_ptr<T>`, where skipping the destructor will leak a refcount to an entire `T` object (thus leak an arbitrary resource, escalating a simple "memory leak" into a "deadlock" or "address in use" or "out of disk space").

Above, you talk about "allowing the users to specialize the is_trivially_moveable value [...] where the OOBE [out-of-box experience] is wrong/buggy." I think it's really dangerous to even _permit_ a state where the out-of-box experience is buggy! Your trait should be _correct by default_, even at the cost of performance. That means never relying on "heuristics" such as "nothrow move-assignable probably means...".  Obviously it's your library and you can do what you want, but _IMHO_ it's a _very_ bad idea to code in heuristics that you _know_ have false positives, and hope that the client programmer can work around them.

The conservative (i.e. always-correct-even-at-the-cost-of-performance) thing to do here is:

    static constexpr bool moved_out_value_is_trivially_destructible = 
      std::is_trivially_destructible_v<T>;

This is _very_ conservative, but I can't think of anything more permissive that isn't also demonstrably wrong and dangerous.

psiha replied:
Thanks for the further engagement (renamed the trait in question, liked your name better:).
So - sure a final, public/wide release should default to correctness (or at least heavily lean so) - however, even though the    rest of the library is pretty mature, this part - containers - is rather new and in the rapid development/tracer-bullet stage, it is currently used in one project which I lead and the developers have direct contact with me - so getting efficiency (speed and binary size-wise) by default while 'gathering knowledge' (by catching bugs with 'strange' types) and 'pushing the envelope' (i.e. creating use cases for the new features we wish to have in the language) is kind of the ('sweet') spot I decided on (for this stage - later on I plan on adding a compiletime switch to choose between strict safety and 'optimistic heuristics' or going strict right away)...

> and hope that the client programmer can work around them.

Just specializing the trait for offending types should be enough (relying on sanitizers and tests to catch those).

psiha replied:
I'll be merging this PR at some point (to bring in the 'too optimistic heuristic' fixes:), feel welcome to comment further/open more issues...

Quuxplusone replied:
> (liked your name better:)

Heh. I thought I had blindly copy-pasted _your_ name, re `moved_out_value_is_trivially_destructible`! I was actually coming to the conclusion myself that the correct name for this property is `(is_)trivially_destructible_after_move(_v)`, and thought I had _you_ to thank for _that!_

To be clear, this property resembles neither [P1144 `is_trivially_relocatable`](https://open-std.org/jtc1/sc22/wg21/docs/papers/2024/p1144r12.html) nor [P2786 `is_trivially_relocatable`](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p2786r11.html); instead, it closely resembles [P1029R2 `move_relocates`](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1029r2.pdf). I think `trivially_destructible_after_move` is a good name for this (P1029-like) property.

Quuxplusone replied:
Argh. I see now that you have _two_ traits (currently named `is_trivially_moveable` and `moved_out_value_is_trivially_destructible`): the first resembles P1144/P2786, and the second resembles P1029R2.

In your block comment you write:

> This library/the author leans toward 'trivially relocatable' to mean that an object can be simply 'paused' (i.e. not a thread safe operation), 'picked up' from its current location, 'droped' at a different location and simply 'resumed' like nothing happened (in the sense of program correctness), IOW an obj can have the value of its this pointer changed w/o violating any of its invariants or preconditions.

Suppose you have _two_ objects of type `O`, and you want to swap their values:
```
int i = 1, j = 2;
O oi = {i}, oj = {j};
swap(oi, oj);
```
For trivially relocatable types `O`, I think you'd agree, this can be accomplished by "pausing" both objects, "picking them both up," and "dropping" them back into each other's locations: `oi` gets the bytes formerly in `oj`, and vice versa. Since they're trivially relocatable, this is fine: the value of each object's `this` pointer changes, that's all.

However, if `O` is `tuple<int&>`, then this simple "pick up and drop" algorithm does _not_ work. I mean it has the wrong behavior for `tuple<int&>` specifically, because `swap(ti, tj)` is documented to swap the bytes of `ti`'s and `tj`'s _pointees_, not the bytes of the `tuple` objects themselves. Q.E.D., logically, `tuple<int&>` must not be trivially relocatable.

P1144 agrees with that conclusion, and provides a formal spec explaining why `tuple<int&>` isn't P1144-trivially-relocatable (specifically, it has a funky assignment operator, which disables P1144-trivial-relocatability). On the other hand, P2786 rejects the conclusion: P2786 _wants_ `tuple<int&>` to be P2786-trivially-relocatable. To preserve the syllogism, P2786 also rejects its major premise: your "pick up and drop" model of trivial relocatability. Under P2786, it's safe to "pick up and drop" a P2786-trivially-relocatable object's value only if the destination is "blank," as in vector reallocation, where you're aiming into a completely new buffer. It's not safe to use P2786-relocation to implement swap, or rotate, or vector::erase, or any other algorithm involving "simultaneous movement" of objects into _each other's_ former positions.

---

The big block comment added in PR #34 talks mainly about "trivial relocatability," and is attached to your trait currently named `is_trivially_moveable`. That trait is used like this in `make_space_for_insert`:

        if constexpr ( is_trivially_moveable<value_type> )
        {
            // does not use is_trivially_moveable/trivial_abi and is incorrect
            // (i.e. an uninitialized_move_backwards is required)
            //std::uninitialized_move_n( &data[ position_index ], elements_to_move, &data[ position_index + n ] );
            std::memmove( &data[ position_index + n ], &data[ position_index ], elements_to_move * sizeof( *data ) );
        }

This memmove does _not_ write only into "blank" space, and so the memmove optimization preserves correctness only when the objects being relocated are _not_ things like `tuple<int&>` or `struct S { pmr::vector<int> v; }`. In P1144, that's spelled:

        if constexpr ( std::is_trivially_relocatable_v<value_type> )

In P2786R11 (which splits P1144's definition of "trivial relocatability" into a pair of traits), that's spelled:

        if constexpr ( std::is_trivially_relocatable_v<value_type> && std::is_replaceable_v<value_type> )

So what you want for your `is_trivially_moveable` trait is:

    template <class T>
    constexpr bool has_property_p_v =
    #if USE_P1144
      std::is_trivially_relocatable_v<T>;  // P1144 semantics
    #elif USE_P2786
      std::is_trivially_relocatable_v<T> &&  // P2786 semantics
      std::is_replaceable_v<T>;
    #else
      std::is_trivially_copyable_v<T>;
    #endif

(2) Currently you constrain your `vm_vector` to accept only `is_trivially_moveable` types. IIUC, your `vm_vector` is a shared-memory container similar to Boost.Interprocess ([like this](https://quuxplusone.github.io/blog/2024/08/23/boost-interprocess-tutorial/)). _If_ that's the case, then it would be unsafe to make a `vm_vector<std::vector<int>>`, is that right? And yet `is_trivially_moveable<std::vector<int>> == true`. Contrariwise, it would be safe to make a `vm_vector<bip::offset_ptr<int>>`, and yet `is_trivially_moveable<bip::offset_ptr<int>> == false`. I don't have a silver-bullet solution to offer you here; all I can suggest is that the constraint is not pulling its weight, and suggest that you leave `vm_vector` unconstrained.  (But if I'm wrong about its purpose, then please ignore this comment.)

(3) Orthogonally, you have your second trait. It's currently named `moved_out_value_is_trivially_destructible`. I suggest naming it `(is_)trivially_destructible_after_move(_v)`. This trait resembles [P1029R2 `move_relocates`](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1029r2.pdf). It has nothing to do with "picking up and dropping"; it does not enable any optimizations involving memcpy/memmove. Instead, _this_ trait enables skipping the destructor after a move. MSVC's sentinel-node `list` is "trivially relocatable" but not "trivially destructible after move"; libc++'s dummy-node `list` is "trivially destructible after move" but not "trivially relocatable." Now that I have it straight in my brain — that you have _two_ traits, with different meanings — I think I have no problem with this _second_ trait. :)

psiha replied:
> However, if O is tuple<int&>, then this simple "pick up and drop" algorithm does not work. I mean it has the wrong behavior for tuple<int&> specifically, because swap(ti, tj) is documented to swap the bytes of ti's and tj's pointees, not the bytes of the tuple objects themselves. Q.E.D., logically, tuple<int&> must not be trivially relocatable.

As you imply this conclusion is disputed (from the very existence of P2786) - IOW from my/the other/P2786 POV this/your conclusion looks like 'circular QEDing' - IOW what to me is the intuitive, expected and desired behaviour is that 'the laws of the universe are the same for all types of relocation' (std::realloc, new + copy + free, insert, erase, push_front...) - IOW if I don't expect my tuple<T&> to change in any observable way due to a vector resize (which might cause its this value to change) I likewise don't expect it to change due to a an erase (which might also cause its this value to change) - IOW the internal implementation details of a container I put my object in should preferably have no influence over the state of object (e.g. I put my tuple<T&> in a deque and don't want  my program suddenly behaving _differently_ in the sense of producing different results rather just having different performance characteristics).
I actually already covered this rationale somewhat in the introductory comment for is_trivially_moveable...
Even for more elaborate/complex yet real cases like pmr containers: yes I'd prefer the more efficient and less surprising pickup&drop (memmove) behaviour rather then try-move-and-throw/or-copy-if-moving-into-a-slot-of-an-empty/moved-out-container-yet-which-has-an-incompatible-allocator.

psiha replied:
> Under P2786, it's safe to "pick up and drop" a P2786-trivially-relocatable object's value only if the destination is "blank," as in vector reallocation, where you're aiming into a completely new buffer. It's not safe to use P2786-relocation to implement swap, or rotate, or vector::erase, or any other algorithm involving "simultaneous movement" of objects into each other's former positions.

True - I'm pushing even beyond P2786 (i.e. I'm in the camp that has no problem with breaking things in order to improve 'the future'...no one forces anyone to move to the next standard, P1881 etc) - backwards compatibility may have been the strength of C++ but, at this stage, it will also be its death (e.g. the stories of players like Google and Apple).

In any case, if pressed to choose between the two proposals I'd go with P2786/two traits - as it gives more optimization opportunities (i.e. at least some room for types that fall under just one of the traits - and thus would be skipped altogether under P1144).

psiha replied:
> This memmove does not write only into "blank" space, and so the memmove optimization preserves correctness only when the objects being relocated are not things like tuple<int&> or struct S { pmr::vector<int> v; }. In P1144, that's spelled:

I take a different view of _correctness_ and _safety_ - i.e. I see nothing unsafe or incorrect with replanting/reseating a tuple<T&> instead of calling move assignment (in cases where move assignment is not explicitly the thing a user wants/'spelled out in code') - it is not like the program is going to crash/enter UB teritory because of it.

psiha replied:
> (2) Currently you constrain your vm_vector to accept only is_trivially_moveable types. IIUC, your vm_vector is a shared-memory container similar to Boost.Interprocess ([like this](https://quuxplusone.github.io/blog/2024/08/23/boost-interprocess-tutorial/)). If that's the case, then it would be unsafe to make a vm_vector<std::vector<int>>, is that right? And yet is_trivially_moveable<std::vector<int>> == true. Contrariwise, it would be safe to make a vm_vector<bip::offset_ptr<int>>, and yet is_trivially_moveable<bip::offset_ptr<int>> == false. I don't have a silver-bullet solution to offer you here; all I can suggest is that the constraint is not pulling its weight, and suggest that you leave vm_vector unconstrained. (But if I'm wrong about its purpose, then please ignore this comment.)

Yes, it is similar - although for persistence (DB engine) not for IPC and yes, I know the two so far mentioned traits do not cover the 'safe to persist' question - I [explicitly mention that](https://github.com/psiha/vm/blob/a19f46df8ed3eebb0b6ffb7f7230466e23b12ce7/include/psi/vm/containers/is_trivially_moveable.hpp#L61) :)
I just got around to that - will commit shortly a new trait, 'does_not_hold_absolute_addresses', related to that...

ps. vm_vector allows for either RAM/pagefile/'temporal storage' or disk/'persistable storage' to be selected at runtime as the backing store (so the extra/third trait is needed only for the persistable storage case.

psiha replied:
> (3) Orthogonally, you have your second trait. It's currently named moved_out_value_is_trivially_destructible. I suggest naming it (is_)trivially_destructible_after_move(_v). This trait resembles [P1029R2 move_relocates](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1029r2.pdf). It has nothing to do with "picking up and dropping"; it does not enable any optimizations involving memcpy/memmove. Instead, this trait enables skipping the destructor after a move. MSVC's sentinel-node list is "trivially relocatable" but not "trivially destructible after move"; libc++'s dummy-node list is "trivially destructible after move" but not "trivially relocatable."

I renamed the trait to trivially_destructible_after_move_assignment - thanks to your input/blog post bringing into focus types which 'exchange' storage in move assignment 👍
(obviously two separate traits are needed...this is quickly getting out of hand 🙈)

psiha replied:
> I just got around to that - will commit shortly a new trait, 'does_not_hold_absolute_addresses', related to that...

@Quuxplusone here it is [_does_not_hold_absolute_addresses_](https://github.com/psiha/vm/pull/34/commits/15ec098c4223d8b18a2119e269627b0098d5233a), of course, feel free to name it better :)

ps. I see you are active in SG14 - I used to follow it more closely - that is also a possible venue for further discussion/exploration.
In short, what we have here so far is:
- trivial relocatability
- replacability
- move-construct implies destruction
- move-assignment implies destruction
- persistability
- yet another type of generic reference (and/or expanded universal reference) [that collapses to pass-by-value for const references to types passable through registers](https://github.com/psiha/vm/blob/master/include/psi/vm/containers/abi.hpp) (or any other reason the compiler deduces it be more efficient to pass-by-val, _regardless of the call context/site_) - IOW a 'reference'/'decorator' which says "I simply need a value - and I do not need the address it comes from and will not change it" (therefore it makes no difference semantically whether you give me a _T const &_ or _T_)

Once again, thanks for chiming in!

Quuxplusone replied:
Yep, I think I agree with your taxonomy (that no two of those six traits are identical). Your "persistability" equals "does not hold absolute addresses," right?
[My "Trivially Relocatable" talk from C++Now 2019](https://www.youtube.com/watch?v=SGdfPextuAU&t=3829s) goes into the caveats around "persistence" a tiny bit. For the reasons given in that talk, I don't think it's even _theoretically possible_ to get compiler assistance with that trait; you'll rely on manual warrants forever.

Mind you, _in practice_ I think there's no chance to get that whole taxonomy of traits into the actual C++ Standard — and in fact WG21 is well on its way to screwing up the first one, "trivial relocatability," right now 🙁  — so I think you'll always rely on manual warrants for most of these properties just as a practical matter.

psiha replied:
hm, this reply somehow got lost...

> However, if O is tuple<int&>, then this simple "pick up and drop" algorithm does not work. I mean it has the wrong behavior for tuple<int&> specifically, because swap(ti, tj) is documented to swap the bytes of ti's and tj's pointees, not the bytes of the tuple objects themselves. Q.E.D., logically, tuple<int&> must not be trivially relocatable.

As you imply this conclusion is disputed (from the very existence of P2786) - IOW from my/the other/P2786 POV this/your conclusion looks like 'circular QEDing' - IOW what to me is the intuitive, expected and desired behaviour is that 'the laws of the universe are the same for all types of relocation' (std::realloc, new + copy + free, insert, erase, push_front...) - IOW if I don't expect my tuple<T&> to change in any observable way due to a vector resize (which might cause its this value to change) I likewise don't expect it to change due to a an erase (which might also cause its this value to change) - IOW the internal implementation details of a container I put my object in should preferably have no influence over the state of object (e.g. I put my tuple<T&> in a deque and don't want  my program suddenly behaving _differently_ in the sense of producing different results rather just having different performance characteristics).
I actually already covered this rationale somewhat in the introductory comment for is_trivially_moveable...
Even for more elaborate/complex yet real cases like pmr containers: yes I'd prefer the more efficient and less surprising pickup&drop (memmove) behaviour rather then try-move-and-throw/or-copy-if-moving-into-a-slot-of-an-empty/moved-out-container-yet-which-has-an-incompatible-allocator.

psiha replied:
> Yep, I think I agree with your taxonomy (that no two of those six traits are identical). Your "persistability" equals "does not hold absolute addresses," right?

Right.

> [My "Trivially Relocatable" talk from C++Now 2019](https://www.youtube.com/watch?v=SGdfPextuAU&t=3829s) goes into the caveats around "persistence" a tiny bit. For the reasons given in that talk, I don't think it's even _theoretically possible_ to get compiler assistance with that trait; you'll rely on manual warrants forever.

Thanks, I'll have a look (I may have viewed it already but even so obviously need a refresher:).


> Mind you, _in practice_ I think there's no chance to get that whole taxonomy of traits into the actual C++ Standard — and in fact WG21 is well on its way to screwing up the first one, "trivial relocatability," right now 🙁 — so I think you'll always rely on manual warrants for most of these properties just as a practical matter.

I would suppose so (one can tell my 'optimism' about the WG21 from a mile away :D) 🤷

psiha replied:
> [My "Trivially Relocatable" talk from C++Now 2019](https://www.youtube.com/watch?v=SGdfPextuAU&t=3829s) goes into the caveats around "persistence" a tiny bit. For the reasons given in that talk, I don't think it's even theoretically possible to get compiler assistance with that trait; you'll rely on manual warrants forever.

Hm - I didn't catch any details mentioning/implying something on why this ('persistability') could not even theoretically be implemented or at least helped by the compiler?
AFAICT the only (even though a very strong and limiting one) requirement for a persistable object is that it contains no addresses (even relative ones, which are in the end based on an absolute one, e.g. bip::offset_ptr - I accordingly renamed the trait to just does_not_hold_addresses) and it seems that this could be automatically (viraly) propagated by the compiler, just like [[trivially_relocatable]] (except for cases where the programmer stores uintptrs which are subsequently reinterpret_casted to pointers).

psiha replied:
> Yep, I think I agree with your taxonomy (that no two of those six traits are identical). Your "persistability" equals "does not hold absolute addresses," right?
[My "Trivially Relocatable" talk from C++Now 2019](https://www.youtube.com/watch?v=SGdfPextuAU&t=3829s)

Additionally - @Quuxplusone your mention of P0023/relocator (the new constructor type/syntax) & N4158/destructive move (the customization point function in particular) reminded me of a similar thing that I already do (although not in the VM library but in the closed source parts that use the lib) and that is sort of a blend of both approaches and which I'd like to add to this 'taxonomy' or (real world) wishlist as it would probably bring the percentage of 'almost' trivially moveable types to '99.999%': so, for types which hold self-references or which manage sub/child objects (which hold backreferences to the parent object) - the move operations cannot be trivial because of the fixups required to the self-references or to the references in sub/child objects - _**but**_ - these fixups are noexcept operations and if we separate them out to a dedicated operator/special function then the compiler and/or the standard library can still for example expand a vector of such types with realloc and afterwards call the fixups on all the objects.
I say it is a mix of the two proposalas because it sort of has the 'postprocess function' of N4158 and a special member function of P0023.
e.g. classes gain an overloadable operator .*>>( ptrdiff_t offset )- which is a noop by default and which recieves the number of bytes the object's this pointer was shifted 'to the right'/'increased'

psiha replied:
'Finally' 🙈 😅, squeezing every last bit of this would be to allow for hybrid default-generated special functions for tuples/classes with multiple members of different types with different characteristics WRT all of the mentioned/proposed traits - e.g. if a class has 5 members of 5 types and if only one of those types is not trivially relocatable but gives the 'move-assignment implies destruction' warrant then the compiler 'should' use memove for all the 4 types and at least skip the destructor for the 5th one in contexts where relocation is requested (instead of just falling back on move-assignment for the entire object).

Quuxplusone replied:
> Hm - I didn't catch any details mentioning/implying something on why this ('persistability') could not even theoretically be implemented or at least helped by the compiler?
> AFAICT the only (even though a very strong and limiting one) requirement for a persistable object is that it contains no addresses (even relative ones, which are in the end based on an absolute one, e.g. bip::offset_ptr [...]

`offset_ptr` is exactly the stumbling block. A persistable object _can_ contain `offset_ptr`s into _itself_; in fact that's the typical case IMHO (like in my blog post on Boost.Interprocess). But it cannot contain `offset_ptr`s into any _other_ object. Unfortunately, the question "Does this `offset_ptr` point into _this_ object or into another object?" cannot be answered at compile time; it's a semantic property of the value of the pointer, not a syntactic property of the type of the pointer.  (And notice that by "into itself" I really mean "into the arena of which it is a part"; e.g. it's OK for a `shm::vector` to contain an `offset_ptr` to an allocation in the same shared-memory segment as the `shm::vector` itself.)

Reductio ad absurdum: A persistable object generally can't contain raw `T*`s at all — neither into itself, nor into any other object. But actually it _can_ contain a `T*` member, as long as that member has the value `NULL`. Persisting the value `NULL` is fine and does not trash it.

So persistability is _in general_ a semantic, dynamic, run-time property. We _almost always_ build persistable objects out of smaller building blocks that are not themselves persistable in isolation.

Quuxplusone replied:
FWIW, your idea of a "postprocess" or "fixup" or "repair" function, to repair the effects of memcpy, is _not_ similar to N4158 as I understand it. N4158 was still just a general relocate operation, like the modern [P2785 "Relocating prvalues"](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2785r3.html) (now sadly abandoned due to the primary author's death) — it allowed you to override the entire `destructive_move` customization point, but that's all; there was no notion of augmenting memcpy with a "repair" function.

Your idea _is_ similar to (and your idea is much better than) ETLCPP's "repair" function: https://www.etlcpp.com/vector.html

psiha committed:
moved_out_value_is_trivially_destructible:
 - renamed to trivially_destructible_after_move
 - extracted it into a standalone/namespace level trait
 - fixed it for MSVC's std::list.
Minor stylistic changes.

psiha committed:
Implemented ra_full_node_iterator equality operator directly instead of relying on the stl_interfaces::iterator_interface provided one (which produces UBSAN noise).
Minor related refactoring and stylistic changes.

Quuxplusone commented on include/psi/vm/containers/vector_impl.hpp:
```diff
@@ -6,8 +6,10 @@
-/// e.g. https://github.com/AmadeusITGroup/amc
+/// TODO extract the entire container subdirectory into a separate library or
+/// merge with prior art
+///  https://github.com/arturbac/small_vectors
+///  https://github.com/AmadeusITGroup/amc
```
FYI: I talked to @arturbac about small_vectors just the other day, so I know that small_vectors is prior art for the `trivially_destructible_after_move` trait (and _not_ for any flavor of `is_trivially_relocatable`).
I talked to @sjanel about AMC longer ago. My understanding is that AMC is prior art for P1144-style `is_trivially_relocatable` and _not_ for `trivially_destructible_after_move`.

psiha replied:
> FYI: I talked to @arturbac about small_vectors just the other day, so I know that small_vectors is prior art for the `trivially_destructible_after_move` trait (and _not_ for any flavor of `is_trivially_relocatable`). I talked to @sjanel about AMC longer ago. My understanding is that AMC is prior art for P1144-style `is_trivially_relocatable` and _not_ for `trivially_destructible_after_move`.

The 'prior art' here is meant WRT extra/custom/non-standard/extended vector-like conatiner designs and implementations :)

Specifically - the goal of vector_impl - abstract or extract out all standard (or 'shareable') API boilerplate - so that different container impls basically have to provide/implement only the 'storage API' (which is different/more basic then the allocator concept - all vectors just need some sort of 'contiguous storage' and no need for generic allocation of 'chunks' from such storage...in effect one usually implements a generic allocator on top of such 'contiguous storage blocks' i.e. memory zones).

Quuxplusone commented on include/psi/vm/containers/is_trivially_moveable.hpp:
```diff
@@ -33,35 +33,75 @@ namespace psi::vm
+    // contrived types support below this line
     std::is_trivially_move_constructible_v<T> ||
-    std::is_trivially_copyable_v<T>
+    ( std::is_trivially_copyable_v<T> && std::is_trivially_destructible_v<T> )
```
Nit: Any trivially copyable type _must_ be trivially destructible ([[class.prop]](https://eel.is/c++draft/class.prop#1)).
```suggestion
    std::is_trivially_copyable_v<T>
```

arturbac commented:
IMHO I was thinking about naming IMHO
is_trivially_relocatable - skip move and destruction on from object , do
just memcpy
is_relocatable (you suggested trivially_destructible_after_move) - skip
destruction on from object during relocation after normal move


On Fri, Jan 10, 2025 at 4:46 PM Quuxplusone ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In include/psi/vm/containers/is_trivially_moveable.hpp
> <https://github.com/psiha/vm/pull/34#discussion_r1910557293>:
>
> >      std::is_trivially_move_constructible_v<T> ||
> -    std::is_trivially_copyable_v<T>
> +    ( std::is_trivially_copyable_v<T> && std::is_trivially_destructible_v<T> )
>
> Nit: Any trivially copyable type *must* be trivially destructible (
> [class.prop] <https://eel.is/c++draft/class.prop#1>).
> ⬇️ Suggested change
>
> -    ( std::is_trivially_copyable_v<T> && std::is_trivially_destructible_v<T> )
> +    std::is_trivially_copyable_v<T>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/psiha/vm/pull/34#pullrequestreview-2542763278>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADSIGYRIPJZR2NPYEZLONMT2J7TMJAVCNFSM6AAAAABU4FE5RCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKNBSG43DGMRXHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Quuxplusone commented:
> IMHO I was thinking about naming IMHO `is_trivially_relocatable` - skip move and destruction on from object , do just memcpy `is_relocatable` (you suggested `trivially_destructible_after_move`) - skip destruction on from object during relocation after normal move

IMO the problem with that naming is that you will then have a large set of real-world types that are `is_trivially_relocatable` (i.e. can be relocated via memcpy) but _not_ `is_relocatable` (i.e. cannot be relocated via their move constructor); for example, Microsoft's `std::list`. I think that indicates that the name is wrong: logically, every type that is bazly fooable should be fooable, even when foo=relocat and baz=trivial. I have a new blog post on this topic here: https://quuxplusone.github.io/blog/2025/01/10/trivially-destructible-after-move/ Particularly notice the caveats at the very end of the post: perhaps `trivially_relocatable_after_move` is still too vague, and it should be `trivially_relocatable_after_move_construction` instead (as opposed to `trivially_relocatable_after_move_assignment`).

psiha committed:
trivially_destructible_after_move:
 - renamed to trivially_destructible_after_move_assignment
 - added a specialization for libstdc++ string.
Expanded/updated comments.
Minor stylistic cleanups.

psiha committed:
static_vector:
 - fixed bugs in assignment operators
 - added fixed_sized_move for more optimization opportunities
 - added is_trivially_moveable specialization to help with contexts/compilers where [[trivial_abi]] isn't enough.

psiha committed:
Added the does_not_hold_absolute_addresses trait (used to constrain vm_vector::map_file).
Minor is_trivially_moveable cleanups.

psiha committed:
Included enums in does_not_hold_absolute_addresses.

psiha committed:
MSVC/ClangCL: fixed trivially_destructible_after_move_assignment specialization for std::list.

psiha committed:
Added a todo note.

psiha committed:
Minor stylistic fix.

psiha committed:
OSX: workarounds for UBSan noise.
Minor other refactoring.

psiha committed:
Renamed does_not_hold_absolute_addresses to does_not_hold_addresses (and expanded the documenting comment for explanation).
Added more 'prior art' links for is_trivially_moveable.
